File: starlore-py/app/services/project_service.py
"""项目 CRUD 服务。"""

from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.exceptions import NotFoundException, UnauthorizedException
from app.models.project import Project
from app.schemas.project import ProjectInfo, ProjectRequest


def _to_info(project: Project) -> ProjectInfo:
    return ProjectInfo(
        id=project.id,
        name=project.name,
        description=project.description,
        url=project.url,
        image=project.image,
        sortOrder=project.sort_order,
        createdAt=project.created_at,
        updatedAt=project.updated_at,
    )


async def list_projects(db: AsyncSession, user_id: int) -> list[ProjectInfo]:
    result = await db.execute(
        select(Project).where(Project.user_id == user_id)
        .order_by(Project.sort_order.asc(), Project.created_at.desc())
    )
    return [_to_info(p) for p in result.scalars().all()]


async def create_project(db: AsyncSession, user_id: int, req: ProjectRequest) -> ProjectInfo:
    now = datetime.now()
    project = Project(
        user_id=user_id,
        name=req.name or "",
        description=req.description,
        url=req.url,
        image=req.image,
        sort_order=req.sortOrder or 0,
        created_at=now,
        updated_at=now,
    )
    db.add(project)
    await db.flush()
    return _to_info(project)


async def update_project(db: AsyncSession, user_id: int, project_id: int, req: ProjectRequest) -> ProjectInfo:
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if project is None:
        raise NotFoundException("项目不存在")
    if project.user_id != user_id:
        raise UnauthorizedException("无权操作该项目")

    if req.name is not None:
        project.name = req.name
    if req.description is not None:
        project.description = req.description
    if req.url is not None:
        project.url = req.url
    if req.image is not None:
        project.image = req.image
    if req.sortOrder is not None:
        project.sort_order = req.sortOrder

    project.updated_at = datetime.now()
    await db.flush()
    return _to_info(project)


async def delete_project(db: AsyncSession, user_id: int, project_id: int) -> None:
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if project is None:
        raise NotFoundException("项目不存在")
    if project.user_id != user_id:
        raise UnauthorizedException("无权操作该项目")

    await db.delete(project)
    await db.flush()

File: starlore-go/internal/services/project_service.go
package services

import (
	"net/http"
	"time"

	"starlore-go/internal/database"
	"starlore-go/internal/models"
	"starlore-go/internal/schemas"
	"starlore-go/internal/utils"

	"github.com/gin-gonic/gin"
)

func ListProjects(c *gin.Context) {
	uid := utils.GetUserID(c)
	var projects []models.Project
	database.DB.Where("user_id = ?", uid).Order("sort_order ASC, created_at DESC").Find(&projects)

	items := make([]schemas.ProjectInfo, len(projects))
	for i, p := range projects {
		items[i] = schemas.ProjectInfo{
			ID:          p.ID,
			Name:        p.Name,
			Description: p.Description,
			URL:         p.URL,
			Image:       p.Image,
			SortOrder:   p.SortOrder,
			CreatedAt:   p.CreatedAt,
			UpdatedAt:   p.UpdatedAt,
		}
	}
	c.JSON(http.StatusOK, gin.H{"data": items})
}

func CreateProject(c *gin.Context, req schemas.ProjectRequest) {
	uid := utils.GetUserID(c)
	now := time.Now()
	project := models.Project{
		UserID:    uid,
		CreatedAt: &now,
		UpdatedAt: &now,
	}
	if req.Name != nil {
		project.Name = *req.Name
	}
	if req.Description != nil {
		project.Description = *req.Description
	}
	if req.URL != nil {
		project.URL = *req.URL
	}
	if req.Image != nil {
		project.Image = *req.Image
	}
	if req.SortOrder != nil {
		project.SortOrder = *req.SortOrder
	}
	database.DB.Create(&project)
	utils.Ok(c, "项目创建成功", gin.H{"id": project.ID})
}

func UpdateProject(c *gin.Context, req schemas.ProjectRequest) {
	uid := utils.GetUserID(c)
	projectID := utils.ParseInt(c.Param("id"), 0)

	var project models.Project
	if err := database.DB.First(&project, projectID).Error; err != nil {
		utils.Fail(c, http.StatusNotFound, "项目不存在")
		return
	}
	if project.UserID != uid {
		utils.Fail(c, http.StatusUnauthorized, "无权操作该项目")
		return
	}

	updates := map[string]interface{}{}
	if req.Name != nil {
		updates["name"] = *req.Name
	}
	if req.Description != nil {
		updates["description"] = *req.Description
	}
	if req.URL != nil {
		updates["url"] = *req.URL
	}
	if req.Image != nil {
		updates["image"] = *req.Image
	}
	if req.SortOrder != nil {
		updates["sort_order"] = *req.SortOrder
	}
	now := time.Now()
	updates["updated_at"] = &now
	database.DB.Model(&project).Updates(updates)
	utils.Ok(c, "项目更新成功", nil)
}

func DeleteProject(c *gin.Context) {
	uid := utils.GetUserID(c)
	projectID := utils.ParseInt(c.Param("id"), 0)

	var project models.Project
	if err := database.DB.First(&project, projectID).Error; err != nil {
		utils.Fail(c, http.StatusNotFound, "项目不存在")
		return
	}
	if project.UserID != uid {
		utils.Fail(c, http.StatusUnauthorized, "无权操作该项目")
		return
	}
	database.DB.Delete(&project)
	utils.Ok(c, "项目删除成功", nil)
}
